### PR TITLE
refactor(staker): staker

### PR DESCRIPTION
**Pull request by notJoon**
# Description

- Remove unused internal functions and imports
- Moved external incentive related functions from the `staker` file to a separate `staker_external_incentive` file
- Extract condition check as functions
- Change the parameters of the `getTokenPairBalanceFromPosition` function
    - The pool path is already obtained before this function is called with `pn.PositionGetPositionPoolKey(tokenId)` method, it is unnecessary to process it internally.
```diff
- func getTokenPairBalanceFromPosition(tokenId uint64) (string, string) {
+ func getTokenPairBalanceFromPosition(poolPath string, tokenId uint64) (string, string) {
- 	poolKey := pn.PositionGetPositionPoolKey(tokenId)
        // ...
}
```

## Newly Added Test Files

- `manage_pool_tiers`
- `reward_internal_emission`
- `reward_manager`
- `reward_recipient_store`
- `staker`

## Test Result

Screenshot 2024-12-16 at 7 32 14 PM

### CODE REVIEW

**Review by r3v4s — CHANGES_REQUESTED** [review]
- tc doesn't run with current master branch as base
- do not inline `get -> data check -> if nil panic -> if not nil type assertion` for avl data
  - make it as separate function
- for testcase screenshot, run it from gnoswap's repo(with current br branch), not from gno

**Comment by notJoon** [comment]
>     * for testcase screenshot, run it from gnoswap's repo(with current br branch), not from gno

I don't think this will be a problem, since I just moved the entire main branch over to the gno side.

File: staker/calculate_pool_position_reward.gno [hunk 1]
```diff
@@ -518,3 +517,13 @@ func externalDepositGnsAmount() uint64 {
 
 	return amount
 }
+
+
+func getPoolTiers() map[string]InternalTier {
```
Comment (r3v4s) [inline staker/calculate_pool_position_reward.gno:522]: should be avl

File: staker/staker.gno [hunk 2]
```diff
@@ -71,310 +86,142 @@ func StakeToken(tokenId uint64) (string, string, string) {
...
-	deposit.numberOfStakes++
-	deposit.stakeTimestamp = time.Now().Unix()
-	deposit.stakeHeight = std.GetHeight()
-	deposit.targetPoolPath = poolPath
-	deposits[tokenId] = deposit
-
-	if callerIsOwner { // if caller is owner, transfer NFT ownership to staker contract
-		transferDeposit(tokenId, consts.STAKER_ADDR)
+	if err := applyStake(result); err != nil {
+		panic(err)
```
Comment (r3v4s) [inline staker/staker.gno:98]: no need err codes?

File: staker/_RPC_api_stake.gno [hunk 3]
```diff
@@ -98,30 +99,38 @@ func ApiGetRewards() string {
...
-			stakedOrCreatedAt := max(deposit.stakeTimestamp, incentive.startTimestamp)
+			stakedOrCreatedAt := max(deposit.stakeTimestamp, ictv.startTimestamp)
 			now := time.Now().Unix()
 			if now < stakedOrCreatedAt {
-				continue
+				return
 			}
 
-			externalWarmUpAmount, exist := positionsExternalWarmUpAmount[tokenId][incentiveId]
+			externalWarmUpAmount, exist := positionsExternalWarmUpAmount[tokenId][ictvId]
```
Comment (r3v4s) [inline staker/_RPC_api_stake.gno:119]: Separate getter(and setter) will be much better for nested avl datas.

File: staker/reward_recipient_store.gno [hunk 4]
```diff
@@ -195,6 +195,33 @@ func NewPoolLiquidity() *PoolLiquidity {
...
+
+	// nil check for totalLiquidity
+	if p.totalLiquidity == nil {
+		p.totalLiquidity = u256.Zero()
+	}
+
+	// Add position's liquidity to total liquidity
+	positionLiquidity := position.GetLiquidity()
+	if positionLiquidity != nil {
+		p.totalLiquidity.Add(p.totalLiquidity, positionLiquidity)
```
Comment (r3v4s) [inline staker/reward_recipient_store.gno:218]: liquidity is already huge number (that's why we use uint256 for it).
adding u256 + u256 does need overflow check.

File: staker/calculate_pool_position_reward.gno (hunk 1 above)
Comment (notJoon) [reply staker/calculate_pool_position_reward.gno:522]: calc_pool_position_XXX files will be removed. don't need to change

File: staker/_RPC_api_stake.gno (hunk 3 above)
Comment (notJoon) [reply staker/_RPC_api_stake.gno:119]: same as here: https://github.com/gnoswap-labs/gnoswap/pull/437#discussion_r1897793290

that global variable will be removed

File: staker/reward_recipient_store.gno (hunk 4 above)
Comment (notJoon) [reply staker/reward_recipient_store.gno]: https://github.com/gnoswap-labs/gnoswap/pull/437/commits/9a1d977976448874c848a1a4d1a2ff63a03357a3

File: staker/staker.gno (hunk 2 above)
Comment (notJoon) [reply staker/staker.gno:98]: The `applyStake` already returns `errNoPermission` in the `transferDeposit` function, so it does not have to handle it again.

**Comment by notJoon** [comment]
branch has merged with PR #438

**Pull request closed without merging**